producer/writer/message_writer.go

```diff
@@ -0,0 +1,425 @@
+	w.queue.PushBack(msg)
+	w.Unlock()
+
+	w.m.writeNew.Inc(1)
+}
+
+func (w *messageWriterImpl) isValidWriteWithLock(nowNanos int64) bool {
+	if w.cutOffNanos > 0 && nowNanos >= w.cutOffNanos {
+		w.m.writeAfterCutoff.Inc(1)
+		return false
+	}
+	if w.cutOverNanos > 0 && nowNanos < w.cutOverNanos {
+		w.m.writeBeforeCutover.Inc(1)
+		return false
+	}
+	return true
+}
+
+func (w *messageWriterImpl) writeWithLock(m *message, nowNanos int64) {
+	m.IncWriteTimesWithLock()
+	m.IncReads()
+	msg, isValid := m.Marshaler()
+	if !isValid {
+		m.DecReads()
+		return
+	}
+	written := false
+	for _, cw := range w.consumerWriters {
+		if err := cw.Write(msg); err != nil {
+			w.m.oneConsumerWriteError.Inc(1)
+			continue
+		}
+		written = true
+		w.m.writeSuccess.Inc(1)
+		break
+	}
+	m.DecReads()
+
+	if !written {
+		// Could not be written to any consumer, will retry later.
+		w.m.allConsumersWriteError.Inc(1)
+	}
+	m.SetRetryAtNanosWithLock(w.nextRetryNanos(m.WriteTimesWithLock(), nowNanos))
+}
+
+// TODO: make retry time strategy configurable.
+func (w *messageWriterImpl) nextRetryNanos(writeTimes int64, nowNanos int64) int64 {
+	half := w.backOffNanos / 2
+	backOff := half + rand.Int63n(half)
+	retryAtNanos := nowNanos + backOff
+	if writeTimes > 1 {
+		retryAtNanos += (writeTimes - 1) * w.backOffNanos
+	}
+	return retryAtNanos
+}
+
+func (w *messageWriterImpl) Ack(meta metadata) {
+	w.acks.ack(meta)
+}
+
+func (w *messageWriterImpl) Init() {
+	w.wg.Add(1)
+	go func() {
+		w.retryUnacknowledgedForever()
+		w.wg.Done()
+	}()
+}
+
+func (w *messageWriterImpl) retryUnacknowledgedForever() {
+	ticker := time.NewTicker(w.opts.MessageRetryBackoff())
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			w.retryUnacknowledged()
+		case <-w.doneCh:
+			return
+		}
+	}
+}
+
+func (w *messageWriterImpl) retryUnacknowledged() {
+	w.RLock()
+	e := w.queue.Front()
+	l := w.queue.Len()
+	w.RUnlock()
+	w.m.queueSize.Update(float64(l))
+	for e != nil {
+		now := w.nowFn()
+		nowNanos := now.UnixNano()
+		w.Lock()
+		e = w.retryBatchWithLock(e, nowNanos)
+		w.Unlock()
+		w.m.retryBatchLatency.Record(w.nowFn().Sub(now))
+	}
+}
+
+// retryBatchWithLock iterates the message queue with a lock.
+// It returns after visited enough items or the first item
+// to retry so it's more fair with live writes.
+func (w *messageWriterImpl) retryBatchWithLock(
+	start *list.Element,
+	nowNanos int64,
+) *list.Element {
+	var (
+		iterated int
+		next     *list.Element
+	)
+	for e := start; e != nil; e = next {
+		iterated++
+		if iterated > w.opts.MessageRetryBatchSize() {
+			return e
+		}
+		next = e.Next()
+		m := e.Value.(*message)
+		if m.IsDroppedOrAcked() {
+			// Try removing the ack in case the data was dropped rather than acked.
+			w.acks.remove(m.Metadata())
+			w.queue.Remove(e)
+			w.mPool.Put(m)
+			continue
+		}
+		if m.RetryAtNanosWithLock() >= nowNanos {
+			continue
+		}
+
+		w.m.writeRetry.Inc(1)
+		w.writeWithLock(m, nowNanos)
+		return next
```

Comment:
Is the `return` here intentional? Seems it'll return immediately after trying to write one element?

Reply:
Yes, see my comments above the function, the thinking here is to spend about the same amount of time on each retry as a new write, this way allows the new writes to be blocked less time.